RFCs/0009-Fn::For.md

```diff
@@ -0,0 +1,903 @@
+            ImageId: !Ref amiId
+        key: !Sub Instance${amiCounter}
+```
+
+#### Nested Fn::Map
+* Using customized names for index and value is particularly useful when nesting declarations of Fn::Map
+```yaml
+Parameters:
+  InstanceSizes:
+    Type: CommaDelimitedList
+    Default: "m1.small,m1.medium"
+  Ipv6Addresses:
+    Type: CommaDelimitedList
+    Default: "ipv6-1,ipv6-2,ipv6-3"
+Resources:
+  Fn::Merge:
+    - Fn::Map:
+        index: i1
+        value: x1
+        collection: !Ref InstanceSizes
+        key: !Sub "Instance${i1}"
+        fragment:
+          Properties:
+            InstanceType: !Ref x1
+            Ipv6Addresses:
+              - Fn::Map:
+                  index: i2
+                  value: x2
+                  collection: !Ref Ipv6Addresses
+                  fragment:
+                    - Ipv6Address: !Ref x2
+```
+* The above template is equivalent to the below:
+```yaml
+
+Parameters:
+  InstanceSizes:
+    Type: CommaDelimitedList
+    Default: "m1.small,m1.medium"
+  Ipv6Addresses:
+    Type: CommaDelimitedList
+    Default: "ipv6-1,ipv6-2,ipv6-3"
+Resources:
+  Instance0:
+    Type: AWS::EC2::Instance
+    Properties:
+      InstanceType: "m1.small"
+      Ipv6Addresses:
+        - Ipv6Address: "ipv6-1"
+        - Ipv6Address: "ipv6-2"
+        - Ipv6Address: "ipv6-3"
+  Instance1:
+    Type: AWS::EC2::Instance
+    Properties:
+      InstanceType: "m1.medium"
+      Ipv6Addresses:
+        - Ipv6Address: "ipv6-1"
+        - Ipv6Address: "ipv6-2"
+        - Ipv6Address: "ipv6-3"
+```
+
+* Another example of using an outer loop’s variable inside inner loop
+```yaml
+Parameters:
+  Subnets:
+      Type: CommaDelimitedList
+      Default: "subnet1,subnet2"
+  TagValues:
+      Type: CommaDelimitedList
+      Default: "tag1,tag2"
+Resources:
+  Fn::Merge:
+    - Fn::Map:
+        index: i1
+        value: x1
+        collection: !Ref Subnets
+        key: !Sub "Instance${i1}"
+        fragment: 
+          InstanceType: "m1.small"
+          SubnetId: !Ref x1
+          Tags:
+            - Fn::Map:
+                index: i2
+                value: x2
+                collection: !Ref TagValues
+                fragment:
+                  Key: !Sub "${x1}"
+                  Value: !Sub "${x2}"
+```
+
+# Potential Follow-up Features
+
+The following features are out of scope of this RFC. Separate RFC could be created in the future if there is customer demand.
+
+### Indexing
+* You may want to reference a specific object among iteratively created objects in any place within template.
+```yaml
+Parameters:
+   Endpoints:
+     Type: CommaDelimitedList
+     Default: "endpoint1,endpoint2,endpoint3"
+Resources:
+  Topic:
+    Fn::Merge:
+      - Fn::Map:
+          collection: Endpoints
+          fragment:
+            Type: AWS::SNS::Topic
+            Properties:
+              Subscription:
+                - Endpoint: !Ref value
+Outputs:
+  TopicOutput:
+    Value: !Select [1, Topic]
+```
```

Comment:
I'm not sure if this example is clear.

With the current example
- I would assume you'd use `key`, so you create multiple resources
- If you do that and use `!Sub Topic${index}`, you can use `!Ref Topic1` for index-based addressing